modules/ngx_http_lua_upstream/src/ngx_http_lua_upstream_module.c

```diff
@@ -0,0 +1,654 @@
+        return NULL;
+    }
+
+    return &peers->peer[id];
+}
+
+
+static int
+ngx_http_lua_get_peer(lua_State *L, ngx_http_upstream_rr_peer_t *peer,
+    ngx_uint_t id)
+{
+    ngx_uint_t     n;
+
+    n = 8;
+
+#if (nginx_version >= 1009000)
+    n++;
+#endif
+
+    if (peer->down) {
+        n++;
+    }
+
+    if (peer->accessed) {
+        n++;
+    }
+
+    if (peer->checked) {
+        n++;
+    }
+
+    lua_createtable(L, 0, n);
+
+    lua_pushliteral(L, "id");
+    lua_pushinteger(L, (lua_Integer) id);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "name");
+    lua_pushlstring(L, (char *) peer->name.data, peer->name.len);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "weight");
+    lua_pushinteger(L, (lua_Integer) peer->weight);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "current_weight");
+    lua_pushinteger(L, (lua_Integer) peer->current_weight);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "effective_weight");
+    lua_pushinteger(L, (lua_Integer) peer->effective_weight);
+    lua_rawset(L, -3);
+
+#if (nginx_version >= 1009000)
+    lua_pushliteral(L, "conns");
+    lua_pushinteger(L, (lua_Integer) peer->conns);
+    lua_rawset(L, -3);
+#endif
+
+    lua_pushliteral(L, "fails");
+    lua_pushinteger(L, (lua_Integer) peer->fails);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "max_fails");
+    lua_pushinteger(L, (lua_Integer) peer->max_fails);
+    lua_rawset(L, -3);
+
+    lua_pushliteral(L, "fail_timeout");
+    lua_pushinteger(L, (lua_Integer) peer->fail_timeout);
+    lua_rawset(L, -3);
+
+    if (peer->accessed) {
+        lua_pushliteral(L, "accessed");
+        lua_pushinteger(L, (lua_Integer) peer->accessed);
+        lua_rawset(L, -3);
+    }
+
+    if (peer->checked) {
+        lua_pushliteral(L, "checked");
+        lua_pushinteger(L, (lua_Integer) peer->checked);
+        lua_rawset(L, -3);
+    }
+
+    if (peer->down) {
+        lua_pushliteral(L, "down");
+        lua_pushboolean(L, 1);
+        lua_rawset(L, -3);
+    }
+
+    return 0;
+}
+
+
+static ngx_http_upstream_main_conf_t *
+ngx_http_lua_upstream_get_upstream_main_conf(lua_State *L)
+{
+    ngx_http_request_t                   *r;
+
+    r = ngx_http_lua_get_request(L);
+
+    if (r == NULL) {
+        return ngx_http_cycle_get_module_main_conf(ngx_cycle,
+                                                   ngx_http_upstream_module);
+    }
+
+    return ngx_http_get_module_main_conf(r, ngx_http_upstream_module);
+}
+
+
+static ngx_http_upstream_srv_conf_t *
+ngx_http_lua_upstream_find_upstream(lua_State *L, ngx_str_t *host)
+{
+    u_char                               *port;
+    size_t                                len;
+    ngx_int_t                             n;
+    ngx_uint_t                            i;
+    ngx_http_upstream_srv_conf_t        **uscfp, *uscf;
+    ngx_http_upstream_main_conf_t        *umcf;
+
+    umcf = ngx_http_lua_upstream_get_upstream_main_conf(L);
+
+#if (NGX_HTTP_UPSTREAM_RBTREE)
```

Comment:
Optimization of mine.

diff --git a/src/ngx_http_lua_upstream_module.c b/src/ngx_http_lua_upstream_module.c
index 7b23787..c5b4c80 100644
--- a/src/ngx_http_lua_upstream_module.c
+++ b/src/ngx_http_lua_upstream_module.c
@@ -517,10 +517,41 @@ ngx_http_lua_upstream_find_upstream(lua_State *L, ngx_str_t *host)
     ngx_http_upstream_main_conf_t        *umcf;
 
     umcf = ngx_http_lua_upstream_get_upstream_main_conf(L);
+
+#if (NGX_HTTP_UPSTREAM_RBTREE)
+
+    ngx_list_part_t                *part;
+
+    uscf = ngx_http_upstream_rbtree_lookup(umcf, host);
+
+    if (uscf != NULL)
+    {
+        return uscf;
+    }
+
+    part = &umcf->implicit_upstreams.part;
+    uscfp = part->elts;
+
+    for (i = 0; /* void */ ; i++) {
+
+        if (i >= part->nelts) {
+            if (part->next == NULL) {
+                break;
+            }
+
+            part = part->next;
+            uscfp = part->elts;
+            i = 0;
+        }
+
+#else
+
     uscfp = umcf->upstreams.elts;
 
     for (i = 0; i < umcf->upstreams.nelts; i++) {
 
+#endif
+
         uscf = uscfp[i];
 
         if (uscf->host.len == host->len
@@ -542,7 +573,30 @@ ngx_http_lua_upstream_find_upstream(lua_State *L, ngx_str_t *host)
 
         len = port - host->data - 1;
 
+#if (NGX_HTTP_UPSTREAM_RBTREE)
+        ngx_str_t addr;
+        addr.data = host->data;
+        addr.len = len;
+        uscf = ngx_http_upstream_rbtree_lookup(umcf, &addr);
+        if (uscf != NULL && uscf->port
+            && uscf->port == n)
+        {
+            return uscf;
+        }
+        part = &umcf->implicit_upstreams.part;
+        uscfp = part->elts;
+        for (i = 0; /* void */ ; i++) {
+            if (i >= part->nelts) {
+                if (part->next == NULL) {
+                    break;
+                }
+                part = part->next;
+                uscfp = part->elts;
+                i = 0;
+            }
+#else
         for (i = 0; i < umcf->upstreams.nelts; i++) {
+#endif
 
             uscf = uscfp[i];